packages/dapi/lib/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.js

```diff
@@ -0,0 +1,189 @@
+  const blockHeadersProto = new BlockHeaders();
+  blockHeadersProto.setHeadersList(
+    blockHeaders.map((blockHeader) => blockHeader.toBuffer()),
+  );
+
+  const response = new BlockHeadersWithChainLocksResponse();
+  response.setBlockHeaders(blockHeadersProto);
+
+  await call.write(response);
+}
+
+/**
+ * Prepare and send chain lock response
+ *
+ * @param {AcknowledgingWritable} call
+ * @param {ChainLock} chainLock
+ * @returns {Promise<void>}
+ */
+async function sendChainLockResponse(call, chainLock) {
+  const response = new BlockHeadersWithChainLocksResponse();
+  response.setChainLock(chainLock.toBuffer());
+
+  await call.write(response);
+}
+
+/**
+ * @param {getHistoricalBlockHeadersIterator} getHistoricalBlockHeadersIterator
+ * @param {CoreRpcClient} coreAPI
+ * @param {ZmqClient} zmqClient
+ * @param {subscribeToNewBlockHeaders} subscribeToNewBlockHeaders
+ * @return {subscribeToBlockHeadersWithChainLocksHandler}
+ */
+function subscribeToBlockHeadersWithChainLocksHandlerFactory(
+  getHistoricalBlockHeadersIterator,
+  coreAPI,
+  zmqClient,
+  subscribeToNewBlockHeaders,
+) {
+  /**
+   * @typedef subscribeToBlockHeadersWithChainLocksHandler
+   * @param {grpc.ServerWriteableStream<BlockHeadersWithChainLocksRequest>} call
+   */
+  async function subscribeToBlockHeadersWithChainLocksHandler(call) {
+    const { request } = call;
+
+    let fromBlockHash = Buffer.from(request.getFromBlockHash()).toString('hex');
+    const fromBlockHeight = request.getFromBlockHeight();
+    const count = request.getCount();
+
+    const newHeadersRequested = count === 0;
+
+    const acknowledgingCall = new AcknowledgingWritable(call);
+
+    const mediator = new ProcessMediator();
+
+    mediator.on(
+      ProcessMediator.EVENTS.BLOCK_HEADERS,
+      async (blockHeaders) => {
+        await sendBlockHeadersResponse(acknowledgingCall, blockHeaders);
+      },
+    );
+
+    mediator.on(
+      ProcessMediator.EVENTS.CHAIN_LOCK,
+      async (chainLock) => {
+        await sendChainLockResponse(acknowledgingCall, chainLock);
+      },
+    );
+
+    if (newHeadersRequested) {
+      subscribeToNewBlockHeaders(mediator, zmqClient, coreAPI);
+    }
+
+    // If block height is specified instead of block hash, we obtain block hash by block height
+    if (fromBlockHash === '') {
+      if (fromBlockHeight === 0) {
+        throw new InvalidArgumentGrpcError('minimum value for `fromBlockHeight` is 1');
+      }
+
+      // we don't need to check bestBlockHeight because getBlockHash throws
+      // an error in case of wrong height
+      try {
+        fromBlockHash = await coreAPI.getBlockHash(fromBlockHeight);
+      } catch (e) {
+        if (e.code === -8) {
+          // Block height out of range
+          throw new NotFoundGrpcError('fromBlockHeight is bigger than block count');
+        }
+
+        throw e;
+      }
+    }
+
+    let fromBlock;
+
+    try {
+      fromBlock = await coreAPI.getBlock(fromBlockHash);
```

Comment:
We can use block stats here as well later.

Reply:
Added TODO